### PR TITLE
Riverpod for cross signing

### PR DESCRIPTION
**Pull request by bitfriend**
Will implement riverpod for cross signing, so that [session review scene can access it](https://github.com/acterglobal/a3/issues/1240).

### CODE REVIEW

**Comment by gnunicorn** [comment]
@bitfriend can you provide a video cast of it working? And/or provide some steps how to test it?

**Comment by bitfriend** [comment]
Yes, will do.
And I notify I had updated `matrix-rust-sdk` but didn't raise PR to their team repo yet.

**Comment by bitfriend** [comment]
This is the case that element requested verification and element started it.

https://github.com/acterglobal/a3/assets/48786056/2f46b125-c75a-4199-bb2a-c136c3560aa1

**Comment by bitfriend** [comment]
This is the case that element requested verification and acter started it.

https://github.com/acterglobal/a3/assets/48786056/2d6f8b0d-92db-4ad5-bab1-8df4a3010768

**Comment by bitfriend** [comment]
This is the case that this device requested verification and other device started it under both are acter app.

https://github.com/acterglobal/a3/assets/48786056/d63f6592-553b-45dc-bcef-9be139b0435c

**Comment by bitfriend** [comment]
As you see in above video, [button layout issue](https://github.com/acterglobal/a3/issues/1455) was not fixed yet.

**Comment by bitfriend** [comment]
This is scenario that Alice device verifies Bob device.
1. Alice device requests verification to Bob device.
2. Bob device accepts that request and both go to verification start scene.
3. Either Alice or Bob clicks start button and they go to verification key matching scene.
4. If Alice's key array is the same as Bob's key array, they clicks match button. They go to verification done scene and verification is finished.
5. If Alice's key array is not the same as Bob's key array, they clicks mismatch button. They go to verification cancel scene and verification is finished.

File: app/lib/features/cross_signing/pages/verification_request.dart [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    return Container(
+      decoration: BoxDecoration(
+        borderRadius: BorderRadius.circular(15),
+      ),
+      child: Column(
+        mainAxisAlignment: MainAxisAlignment.start,
+        mainAxisSize: MainAxisSize.min,
+        children: [
+          Flexible(
+            flex: isDesktop ? 2 : 1,
```
Comment (gnunicorn) [inline app/lib/features/cross_signing/pages/verification_request.dart:30]: I am not convinced these `isDesktop` calls are that great, as in general we try to avoid that binary view of things: what if we are in a landscape mode on an iPad ... that is closer to the desktop but isn't covered by this flag .... I am okay with keeping this here for now, but in general this isn't a great pattern.

**Review by gnunicorn — APPROVED** [review]
Nothing major from my end, but a few requests for clarification. Also not emerged enough in the details of that process to ensure the flow is absolutely correct myself though...

File: Cargo.toml [hunk 2]
```diff
@@ -8,35 +8,35 @@ members = [
 default-members = ["native/acter"]
 
 [workspace.dependencies.matrix-sdk]
-git = "https://github.com/matrix-org/matrix-rust-sdk"
-rev = "f14c00db82539a2e9135b73568c1f68354c47255"
+git = "https://github.com/acterglobal/matrix-rust-sdk"
+rev = "898e8e149ca35c608fb1787ab312e05e3c2f6a5f"
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply Cargo.toml:12]: Appended some cases of cancel msg [there](https://github.com/acterglobal/matrix-rust-sdk/blob/898e8e149ca35c608fb1787ab312e05e3c2f6a5f/crates/matrix-sdk-crypto/src/verification/mod.rs#L367-L370)

File: app/lib/features/cross_signing/pages/verification_ready.dart [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                  height: 50,
+                  child: CircularProgressIndicator(),
+                ),
+              ),
+            ),
+          ),
+          Flexible(
+            flex: 1,
+            child: TextButton(
+              onPressed: () {},
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/features/cross_signing/pages/verification_ready.dart:78]: This option came from element app.
So I had left this part for the future development.
Then I will comment it.

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bitfriend) [reply Cargo.toml]: Will raise PR to matrix repo?

[the remark this answers is not in the document]
Comment (gnunicorn) [reply Cargo.toml]: yes, please. we always want to stick with matrix-org and only diverge temporarily.

File: app/lib/features/cross_signing/pages/verification_ready.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (gnunicorn) [reply app/lib/features/cross_signing/pages/verification_ready.dart:78]: Future development is good, but buttons that do nothing on click are a bad user experience. We don't to  commit any additional ones ...

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bitfriend) [reply Cargo.toml]: Opened [PR](https://github.com/matrix-org/matrix-rust-sdk/pull/3209).

File: app/pubspec.yaml [hunk 4]
```diff
@@ -125,6 +124,7 @@ dependencies:
   video_player: ^2.8.1
   share_plus: ^7.2.1
   carousel_indicator: ^1.0.6
+  events_emitter: ^0.5.2
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/pubspec.yaml:127]: I defined `init` and `launch` events additionally else matrix-defined events to deal with [verification of other device in session review scene](https://github.com/acterglobal/a3/issues/1240).
`events_emitter` covers both of `init/launch` and `request/ready/start/accept/cancel/key/mac/done`.

**Comment by bitfriend** [comment]
I could remove some codes for enforcement of stage in verification.
For example, the following code:
```dart
          // go to onReady stage and wait for other's ready
          Future.delayed(const Duration(milliseconds: 500), () {
            _verifEmitter.emit('verification.ready', event);
          });
```
To remove this hack code, I installed sas event handler at `onRequestTransitioned()`.

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bitfriend) [reply Cargo.toml]: My mistake closed [that PR](https://github.com/matrix-org/matrix-rust-sdk/pull/3209).
[Another PR](https://github.com/matrix-org/matrix-rust-sdk/pull/3217) was merged to matrix repo.

**Comment by bitfriend** [comment]
Made some changes in localization via fix of conflicts during merge from `main`.

**Comment by bitfriend** [comment]
This is screenshot of acter that I modified for both of mobile & desktop.
image
image
As you see, all rows in dialog are distributed evenly in vertical direction.
The mismatch & match buttons in action bar have the same width and action bar are centered horizontally.

File: app/lib/features/cross_signing/cross_signing.dart [hunk 5]
```diff
@@ -1,874 +1,468 @@
...
-  late StreamSubscription<VerificationEvent>? _verificationPoller;
-  final Map<String, VerificationProcess> _processMap = {};
-  bool _mounted = true;
-  bool isDesktop = Platform.isWindows || Platform.isMacOS || Platform.isLinux;
-
-  CrossSigning({required this.client}) {
-    _installVerificationEvent();
-  }
+  final Client client;
+  final Ref ref;
```
Comment (gnunicorn) [inline app/lib/features/cross_signing/cross_signing.dart:22]: this is never needed.

File: app/lib/features/cross_signing/cross_signing.dart [hunk 6]
```diff
@@ -1,874 +1,468 @@
...
-  bool waitForMatch = false;
-  late StreamSubscription<VerificationEvent>? _verificationPoller;
-  final Map<String, VerificationProcess> _processMap = {};
-  bool _mounted = true;
-  bool isDesktop = Platform.isWindows || Platform.isMacOS || Platform.isLinux;
-
-  CrossSigning({required this.client}) {
-    _installVerificationEvent();
-  }
+  final Client client;
```
Comment (gnunicorn) [inline app/lib/features/cross_signing/cross_signing.dart:21]: what is client ever needed for? Did I miss something?

File: app/lib/features/cross_signing/cross_signing.dart [hunk 7]
```diff
@@ -1,874 +1,468 @@
...
+  String? _flowId; // if not null, verification flying now
+  bool _keyExchanged = false;
+  Stream<VerificationEvent>? _verifListener;
+  StreamSubscription<VerificationEvent>? _verifPoller;
 
-  void _installVerificationEvent() {
-    _verificationPoller = client.verificationEventRx()?.listen((event) {
+  CrossSigning(this.client, this.ref) {
+    _verifListener = client.verificationEventRx(); // keep it resident in memory
+    _verifPoller = _verifListener?.listen((event) {
```
Comment (gnunicorn) [inline app/lib/features/cross_signing/cross_signing.dart:33]: I don't quite understand the point of this `CrossSigning` object and it's `EventEmitter`. Why isn't this just wrapped into a [riverpod `StreamProvider`](https://riverpod.dev/docs/providers/stream_provider)? Seems to much easier, less code, better integrated into our actual system and doesn't need the newly added event_emitter dependency...

File: app/lib/features/cross_signing/cross_signing.dart [hunk 8]
```diff
@@ -1,874 +1,468 @@
...
-        child: _buildOnRequest(context, event, flowId),
-      ),
-      isDismissible: false,
-    );
+  void installEventFilter(BuildContext context) {
+    _verifEmitter.on('verification.init', (_) {
+      onVerificationInit(context);
+    });
+
+    _verifEmitter.on<VerificationEvent>('request.created', (event) {
```
Comment (gnunicorn) [inline app/lib/features/cross_signing/cross_signing.dart:100]: using riverpod stream provider would also make this a bit cleaner as it would split the emission and from the viewer in a cleaner fashion ...

**Review by gnunicorn — CHANGES_REQUESTED** [review]
I didn't review the entire thing as I found its most inner core needs a refactoring into stream providers rather than using event_emitters. So I stopped after there.

File: app/lib/features/chat/models/chat_input_state/chat_input_state.freezed.dart [hunk 9]
```diff
@@ -157,11 +157,11 @@ class _$ChatInputStateCopyWithImpl<$Res, $Val extends ChatInputState>
 }
 
 /// @nodoc
-abstract class _$$_ChatInputStateCopyWith<$Res>
+abstract class _$$ChatInputStateImplCopyWith<$Res>
```
Comment (gtalha07) [inline app/lib/features/chat/models/chat_input_state/chat_input_state.freezed.dart:160]: This seems unrelated changes...

File: app/lib/features/cross_signing/cross_signing.dart (hunk 7 above)
Comment (gtalha07) [reply app/lib/features/cross_signing/cross_signing.dart:33]: Agree, could use ```StreamProvider``` for verification stream here with having ref listener and do appropriate calls based on that...

File: app/lib/features/cross_signing/pages/request_cancelled.dart [hunk 10]
```diff
@@ -0,0 +1,68 @@
+import 'package:acter/common/themes/app_theme.dart';
+import 'package:atlas_icons/atlas_icons.dart';
+import 'package:flutter/material.dart';
+
+class RequestCancelledPage extends StatelessWidget {
```
Comment (gtalha07) [inline app/lib/features/cross_signing/pages/request_cancelled.dart:5]: Nice to have these verification views being separately contained but we they can't be inferred as pages as they don't include ```Scaffold```, I would keep its outer directory as ```widgets/views``` and renaming them to replace ``View``` as ```RequestCancelledView```.

File: app/lib/features/cross_signing/pages/request_created.dart [hunk 11]
```diff
@@ -0,0 +1,64 @@
+import 'package:acter/common/themes/app_theme.dart';
+import 'package:atlas_icons/atlas_icons.dart';
+import 'package:flutter/material.dart';
+
+class RequestCreatedPage extends StatelessWidget {
```
Comment (gtalha07) [inline app/lib/features/cross_signing/pages/request_created.dart:5]: These views seems like most of its inner components are redundant. I would suggest to make one resusable outer wrapper with its inner component as param or strings.

File: app/lib/features/cross_signing/pages/request_ready.dart [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+              padding: EdgeInsets.all(25),
+              child: SizedBox(
+                width: 50,
+                height: 50,
+                child: CircularProgressIndicator(),
+              ),
+            ),
+          ),
+          const Spacer(),
+          // TextButton(
```
Comment (gtalha07) [inline app/lib/features/cross_signing/pages/request_ready.dart:50]: This leaves commented. We can remove it!

File: app/lib/features/cross_signing/pages/sas_cancelled.dart [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    );
+  }
+
+  Widget buildTitleBar(BuildContext context) {
+    // has no close button
+    return Row(
+      children: [
+        Container(
+          padding: const EdgeInsets.all(10),
+          child: Icon(isDesktop ? Atlas.laptop : Atlas.phone),
```
Comment (gtalha07) [inline app/lib/features/cross_signing/pages/sas_cancelled.dart:62]: Same as above comment.

File: app/lib/features/settings/widgets/session_card.dart [hunk 14]
```diff
@@ -59,7 +55,7 @@ class SessionCard extends ConsumerWidget {
                   Padding(
                     padding: const EdgeInsets.symmetric(horizontal: 5),
                     child: Text(
-                      L10n.of(ctx).logOut,
```
Comment (gtalha07) [inline app/lib/features/settings/widgets/session_card.dart:62]: Why is this being unlocalised ? We are replacing all static strings with localisation...

File: app/pubspec.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (gtalha07) [reply app/pubspec.yaml:127]: Still I think with ```StreamProvider``` it'd make it more clean and readable. You'd want to try it.

**Review by gtalha07 — CHANGES_REQUESTED** [review]
It'd be ideal to have entire event emission flow be dealt with ```StreamProvider``` as we are unnecessarily adding another dependency which has minimal use-case but also can be handled alternatively with riverpod. 
Other than that, there are some code style changes and localisation which can be addressed after above :)

File: app/lib/features/cross_signing/cross_signing.dart (hunk 5 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/cross_signing.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/c9c0569cca442daf8549b67deae5e9980bf7d884

File: app/lib/features/cross_signing/cross_signing.dart (hunk 6 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/cross_signing.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/c9c0569cca442daf8549b67deae5e9980bf7d884

File: app/lib/features/cross_signing/cross_signing.dart (hunk 7 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/cross_signing.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/c9c0569cca442daf8549b67deae5e9980bf7d884

File: app/lib/features/cross_signing/cross_signing.dart (hunk 8 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/cross_signing.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/c9c0569cca442daf8549b67deae5e9980bf7d884

File: app/lib/features/settings/widgets/session_card.dart (hunk 14 above)
Comment (bitfriend) [reply app/lib/features/settings/widgets/session_card.dart:62]: Localization for cross signing contained several parametric sentences, like `Waiting for %s…`.
So I introduced `sprintf` package for them before.
But now `arb` for l10n is not used much in acter and we are almost using english sentence as UI text directly.
I thought it is better to remove `sprintf` from dependencies.

File: app/lib/features/cross_signing/pages/request_cancelled.dart (hunk 10 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/pages/request_cancelled.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/edb1022a67a1e0c339a3a546e3522ec47bdcf1e4

File: app/lib/features/settings/widgets/session_card.dart (hunk 14 above)
Comment (bitfriend) [reply app/lib/features/settings/widgets/session_card.dart:62]: When needed, I will restore them to l10n using appropriate manner.

File: app/lib/features/cross_signing/pages/request_created.dart (hunk 11 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/pages/request_created.dart]: Each dialog have different elements.
It is hard to abstract these dialogs into single wrapper.

File: app/lib/features/cross_signing/pages/request_ready.dart (hunk 12 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/pages/request_ready.dart]: This part was left for future implement of qrcode scan.

File: app/lib/features/cross_signing/pages/request_ready.dart [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    );
+  }
+
+  Widget buildTitleBar(BuildContext context) {
+    // has close button
+    return Row(
+      children: [
+        Container(
+          padding: const EdgeInsets.all(10),
+          child: Icon(isDesktop ? Atlas.laptop : Atlas.phone),
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply app/lib/features/cross_signing/pages/request_ready.dart]: Here `isDesktop` is used to show whether current device is pc or mobile.
It isn't used for dimension.

File: app/lib/features/cross_signing/pages/sas_cancelled.dart (hunk 13 above)
Comment (bitfriend) [reply app/lib/features/cross_signing/pages/sas_cancelled.dart]: Here `isDesktop` is used to show whether current device for verification is pc or mobile.
It isn't used for dimension.

File: app/lib/features/home/views/home_body.dart [hunk 16]
```diff
@@ -0,0 +1,528 @@
...
+  ConsumerState<ConsumerStatefulWidget> createState() => HomeBodyState();
+}
+
+class HomeBodyState extends ConsumerState<HomeBody> {
+  final GlobalKey<ScaffoldState> _key = GlobalKey<ScaffoldState>();
+
+  @override
+  void initState() {
+    super.initState();
+  }
```
Comment (gnunicorn) [inline app/lib/features/home/views/home_body.dart:54]: unnecessary.
```suggestion
```

File: app/lib/features/home/views/home_body.dart [hunk 17]
```diff
@@ -0,0 +1,528 @@
...
+                  : null,
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+
+  void onStateChange(VerificationState? prev, VerificationState next) {
+    if (prev == next) return; // avoid too frequent callbacks
```
Comment (gnunicorn) [inline app/lib/features/home/views/home_body.dart:156]: why would that ever happen? Isn't the better way to test for that before setting the `state`? So that we only ever triggered an update if there really was one?

File: app/lib/features/home/views/home_body.dart [hunk 18]
```diff
@@ -0,0 +1,528 @@
...
+      case 'sas.confirmed':
+        onSasConfirmed(next.event!);
+        break;
+      case 'sas.done':
+        onSasDone(next.event!);
+        break;
+    }
+  }
+
+  void onVerificationInit() {
```
Comment (gnunicorn) [inline app/lib/features/home/views/home_body.dart:206]: none of these is technically necessary to be on the shell-widget, isn't it? How about moving them into a separate widget for the cross-singing only then just to clean things clean and separated...

File: app/lib/features/settings/widgets/session_card.dart [hunk 19]
```diff
@@ -150,7 +151,9 @@ class SessionCard extends ConsumerWidget {
     await client.installRequestEventHandler(event.flowId());
 
     // force request.created, because above loop starts from request.ready
-    final crossSigning = ref.read(syncStateProvider.notifier).crossSigning;
-    crossSigning.emitEvent<VerificationEvent>('request.created', event);
+    // final crossSigning = ref.read(syncStateProvider.notifier).crossSigning;
+    // crossSigning.emitEvent<VerificationEvent>('request.created', event);
```
Comment (gnunicorn) [inline app/lib/features/settings/widgets/session_card.dart:155]:
```suggestion
```

File: app/lib/features/home/views/home_body.dart [hunk 20]
```diff
@@ -0,0 +1,528 @@
...
+  void initState() {
+    super.initState();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    ref.listen(verificationStateProvider, onStateChange);
+    final bottomBarNav = ref.watch(bottomBarNavProvider(context));
+    final keyboardVisibility = ref.watch(keyboardVisibleProvider);
+    return CallbackShortcuts(
```
Comment (gnunicorn) [inline app/lib/features/home/views/home_body.dart:61]: I just assumed that all that below here basically didn't change from the place it was copied from ...

File: app/lib/features/home/pages/home_shell.dart [hunk 21]
```diff
@@ -80,11 +72,11 @@ class HomeShellState extends ConsumerState<HomeShell> {
...
-    ref.listen(syncStateProvider, (prev, next) {
-      if (next.initialSync) return;
-      final crossSigning = ref.read(syncStateProvider.notifier).crossSigning;
-      crossSigning.installEventFilter(context);
-    });
+    // ref.listen(syncStateProvider, (prev, next) {
+    //   if (next.initialSync) return;
+    //   final crossSigning = ref.read(syncStateProvider.notifier).crossSigning;
+    //   crossSigning.installEventFilter(context);
+    // });
```
Comment (gnunicorn) [inline app/lib/features/home/pages/home_shell.dart:79]:
```suggestion
```

File: app/lib/features/cross_signing/providers/notifiers/verification_notifiers.dart [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+      /* request_verification_handler: verifier makes request */
+      case 'VerificationRequestState::Created':
+        state = VerificationState(
+          stage: 'request.created',
+          event: event,
+          isVerifier: state.isVerifier,
+          flowId: state.flowId,
+          keysExchanged: state.keysExchanged,
+        );
+        break;
```
Comment (gnunicorn) [inline app/lib/features/cross_signing/providers/notifiers/verification_notifiers.dart:82]: pretty sure @gtalha07 is happy to teach you about `freezed` as well at some point, as it would make this code a lot less verbose... but IMHO this is fine for now ;)

**Review by gnunicorn — CHANGES_REQUESTED** [review]
Great job, mate! You got rid of the extra library and cleaned up the code nicely into `riverpod`. very happy :) Just one more thought about separating the listener out into its own widget either, then this should be good to go :)

File: app/lib/features/home/views/home_body.dart (hunk 16 above)
Comment (bitfriend) [reply app/lib/features/home/views/home_body.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/5164f1e70bb751c241b9ce50e4a38829a0309d14

File: app/lib/features/settings/widgets/session_card.dart (hunk 19 above)
Comment (bitfriend) [reply app/lib/features/settings/widgets/session_card.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/5164f1e70bb751c241b9ce50e4a38829a0309d14

File: app/lib/features/home/pages/home_shell.dart (hunk 21 above)
Comment (bitfriend) [reply app/lib/features/home/pages/home_shell.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/5164f1e70bb751c241b9ce50e4a38829a0309d14

File: app/lib/features/home/views/home_body.dart (hunk 17 above)
Comment (bitfriend) [reply app/lib/features/home/views/home_body.dart]: Fixed by https://github.com/acterglobal/a3/pull/1257/commits/517e209e933acad917d0c878c5758f938638146a

File: app/lib/features/home/views/home_body.dart (hunk 18 above)
Comment (bitfriend) [reply app/lib/features/home/views/home_body.dart:204]: I removed `event_emitter` and made `verificationStateProvider` according to your recommendation.
`ref.listen(verificationStateProvider, onStateChange)` should be called in `build()` of `HomeBody`, before that provider can work properly.
Do you want cross signing widget that has no elements and shows only popup dialog for verification?

File: app/lib/features/cross_signing/providers/notifiers/verification_notifiers.dart [hunk 23]
```diff
@@ -9,35 +9,8 @@ final _log = Logger('a3::cross_signing::notifier');
...
-      return false;
-    }
-    // don't compare isVerifier & flowId & keysExchanged
-    // because they are changed via changeState in ref.listen
-    return true;
-  }
-
-  @override
-  int get hashCode => stage.hashCode;
+  const VerificationState({required this.stage, this.event});
```
Comment (gnunicorn) [inline app/lib/features/cross_signing/providers/notifiers/verification_notifiers.dart:13]: nice reduction of code! very good!

**Review by gnunicorn — APPROVED** [review]
looking good.

File: app/lib/features/settings/widgets/session_card.dart (hunk 14 above)
Comment (bitfriend) [reply app/lib/features/settings/widgets/session_card.dart:64]: Will fix it in [another PR](https://github.com/acterglobal/a3/pull/1539).
